[12.0][MIG] l10n_de_tax_statement

astirpe opened the pull request:
This is the porting to V12 of module `l10n_de_tax_statement` (German VAT Statement)

tv-openbig committed:
module with german vat tax declaration report

we have cloned and adopted the l10n-netherlands tax report module from
Onestein B.V.  The report is constructed in the same way as official
german „Umsatzsteuervoranmeldung“. The values from the report could be
easily transmitted via official portal https://www.elster.de/ in order
to declare and transfer your monthly (or quarterly) vat taxes.

tv-openbig committed:
removal of unnecessary lines

tv-openbig committed:
module german intra-community transactions declaration (Zusammenfassende Meldung)

This module extends the *German VAT Statement* module
(Umsatzsteuervoranmeldung), by adding the statement for the
*Intra-Community transactions declaration* (Zusammenfassende Meldung).

As for the German VAT declaration this module is a cloned adoption from
Onestein’s OCA module 10n_nl_vat_statement_icp.

This report is based on the line *Steuerfr. Umsätze mit Vorsteuerabz.
innerg. Lieferungen (§4 Nr. 1b) an Abnehmer mit USt-ID (41)* and *Nicht
steuerb. sonst. Leist. gem. § 18b S. 1 Nr. 2 (21)* of the German VAT
Statement report.
The period is also the same as the one selected in the VAT statement
report.

This additional VAT declaration report includes:

* the VAT identification numbers of your customers
* the total amount of intra-Community supplies from Germany for every
customer during the selected period
* The values for product deliveries and services are separated in two
columns.
* If you want to transmit the values to the official report form, f.e.
by "www.elster.de" you would have to enter two lines if both columns
product + services contains values.

tv-openbig committed:
removal of unnecessary lines

tv-openbig committed:
mainly fixes to correct pylint and flake errors

tv-openbig committed:
mainly fixes to correct pylint and flake errors

tv-openbig committed:
further pylint flake fixes

tv-openbig committed:
further pylint flake fixes

tv-openbig committed:
fix of pylint and travis errors

************* Module l10n_de_toponyms
l10n_de_toponyms/wizard/l10n_de_toponyms_zipcodes.xml:1:
[W7930(file-not-used), ]  not used from manifest
************* Module l10n_de_tax_statement_zm
l10n_de_tax_statement_zm/readme/USAGE.rst:5: [E7901(rst-syntax-error),
]  Enumerated list ends without a blank line; unexpected unindent.
l10n_de_tax_statement_zm/readme/USAGE.rst:5: [E7901(rst-syntax-error),
]  Inline emphasis start-string without end-string.
l10n_de_tax_statement_zm/readme/USAGE.rst:5: [E7901(rst-syntax-error),
]  Inline emphasis start-string without end-string.
-----------------------------------
Your code has been rated at 9.87/10

tv-openbig committed:
fix of a missing state, removal of unnecessary test files, fix in config wizard

most removed tests are not relevant for germany as they was related in
the clone from netherlands to invoice basis (related to invoice or
account move dates which is probably not relevant for germany). The
change in the config wizard was required as the tax tag mappings for
the tax statement should work with skr03 and skr04 accounting plans.

tv-openbig committed:
Fix of a missing dependency, flake8, testing cases fix

tv-openbig committed:
fix of indention issue

The prior change raised an error. The two additional dependencies
account_invoicing, l10n_de should help to allow immediate runbot
testing without any manual installation steps in advance.

tv-openbig committed:
further fixes and readme documents

tv-openbig committed:
some changes in tests

I have applied proposed changes of Andrea Stirpe in order to get
through the testing procedure.

tv-openbig committed:
some changes in tests

I have applied proposed changes of Andrea Stirpe in order to get
through the testing procedure.

tv-openbig committed:
change of python formula

astirpe committed:
View past invoices moves when country is not Germany

astirpe committed:
Sort past invoices by date

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

oca-travis committed:
[UPD] Update l10n_de_tax_statement.pot

oca-travis committed:
[UPD] Update l10n_de_tax_statement_zm.pot

astirpe committed:
[12.0][MIG] l10n_de_tax_statement

Guenzn commented on l10n_de_tax_statement/README.rst:
```diff
@@ -0,0 +1,129 @@
+
+To install this module, you need to:
+
+#. Install module *l10n_de_tax_statement* version >= *11.0.1.0.0*.
```
should be version >= 12.0.1.0.0

astirpe replied:
Done!

Guenzn commented on l10n_de_tax_statement/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1,8 @@
+
+If the default Odoo German chart of accounts is installed (module l10n_de) you are able to select in the settings if you want to use the skr03 or skr04 chart variant. By installing and configuring your favored german account chart the tax tags from the module l10n_de are automatically present in the database. If this is the case, go to menu: Invoicing -> Configuration -> Accounting -> German Tax  ...
+
+If a non-standard chart of accounts is installed, you have to manually create the tax tags and properly set them into the tax definition. If you create another german account chart (f.e. l10n_de_ikr) you can still depend on l10n_de module in order to benefit from the generic tax tags for germany. If you won't use l10n_de as a base module you have to configure at first your own tax tags. After tha ...
```
I would hint here, that it is required to click Apply the first time!

astirpe replied:
> click Apply to confirm

Should be enough IMO.

Guenzn commented on l10n_de_tax_statement/README.rst:
```diff
@@ -0,0 +1,131 @@
+
+If a non-standard chart of accounts is installed, you have to manually create the tax tags and properly set them into the tax definition. If you create another german account chart (f.e. l10n_de_ikr) you can still depend on l10n_de module in order to benefit from the generic tax tags for germany. If you won't use l10n_de as a base module you have to configure at first your own tax tags. After tha ...
+
+The user must belong to the *Show Full Accounting Features* group, to be able to access the `Invoicing -> Configuration -> Accounting -> NL BTW Tags` menu.
```
Should be "DE" for German localization?

astirpe replied:
Yes, but in this case the name of the menu is: `Ust-Voranmeldung`. Fixed!

astirpe replied:
:thinking: but in other places of the readme file there's written `German Tax Tags` for the same menu. I'm going to set `German Tax Tags` in the menu and update the readme accordingly.

Guenzn replied:
Yeah, menu names needs to be the same in readme as in odoo.  :+1:

astirpe replied:
Ok, should be correct now:
`Invoicing -> Configuration -> Accounting -> German Tax Tags`.

Guenzn commented on l10n_de_tax_statement/models/l10n_de_tax_statement_config.py:
```diff
@@ -0,0 +1,64 @@
+from odoo import fields, models
+
+
+class VatStatementConfig(models.Model):
```
Modelname differs a little bit from class name sometimes....

astirpe replied:
In the [OCA Guidelines](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst) is not mentioned that the class name should be the same as the model name.

Changing the class name in UpperCamelCase would end up in "_L10nDeTaxStatementConfig_". Not so readable IMO. Is it ok to keep the class name as it is?

Guenzn reviewed (commented):
Good but sometimes class and model names are not exactly the same.
Did a code review not a functional

astirpe commented:
Extracted from the document "_2019 Berlin OCA Days Spring Sprint_":


> The module l10n_de_tax_statement_zm is not exactly fitting the german report template of elsteronline.de. As we have cloned and adopted the module from netherlands the report (columns) does fit to netherlands standard. In the netherlands they have two columns to separate intra-community sales towards one partner separating the service sales in one column and the physical goods sales into the second column. 
> In germany we have just column for the invoices amount (for services + goods). Furthermore there is a column where we have to specify if this invoiced amount was a service or a goods invoicing. Here is the original form view:
> 
> https://www.bzst.de/DE/Steuern_International/USt_Kontrollverfahren_ZM_eCommerce/Zusammenfassende_Meldungen/Formulare/ZM_Formular_PDF.pdf?__blob=publicationFile . Here i want to point your interest to column “3” on side 2.

pedrobaeza commented:
Something similar is being discussed in #44

astirpe commented:
@pedrobaeza thanks!
Maybe splitting this PR in two separate PRs, one for each module, would help.

astirpe commented:
Original PR is now split in two: the migration of `l10n_de_tax_statement_zm` (German VAT Statement Extension) is moved into PR #50

haconsult commented:
Hi, 
it seems as if the numbers of fields do not match te statement form on elster-online. Please check e.g. field #53. It is "Umsatzsteuer" but in elster-online it is"Vorsteuerbeträge aus Rechnungen von anderen Unternehmern...". All following numbers do not match.

astirpe commented:
@tv-openbig Would you to take a look at https://github.com/OCA/l10n-germany/pull/39#issuecomment-520252864?

tv-openbig commented:
> Hi,
> it seems as if the numbers of fields do not match te statement form on elster-online. Please check e.g. field #53. It is "Umsatzsteuer" but in elster-online it is"Vorsteuerbeträge aus Rechnungen von anderen Unternehmern...". All following numbers do not match.

Yes indeed if you compare my screenshot you can see on left side the official "2018" version and on the right side the "2019" version. Current version in Odoo fits to 2018 edition of tax declaration if you regard the line numbering.

<img width="1225" alt="Bildschirmfoto 2019-08-12 um 14 54 08" src="https://user-images.githubusercontent.com/8348598/62866694-c1801080-bd11-11e9-8138-788daf7aef85.png">

However much more important in practise as the line numbering are the two digits code numbers formatted in bold letters in front of the base amount / tax amount values. Indeed there was obviously a removal of lines 49 /  51 (with their corresponding code numbers 52,53 & 78,79). In 2019 version they have merged the both lines with their corresponding tax codes under "old" line 52 (code number 84 / 85). 

At the moment i have no idea how we should handle those changes in the official tax 
declarations happening from year to year (deleted lines / deleted or merged tax cases) in our module. Maybe we would have to copy, split & modify the module "l10n_de_tax_statement" into "l10n_de_tax_statement_2018" and "l10n_de_tax_statement_2019", "l10n_de_tax_statement_2020" with separated menus, views.

Furthermore some taxes in the account charts defined to declare the lines 49 / 51 (2018) would be obsolete in 2019 but used in 2018 invoices. In addition the tax code naming in Odoo pointing to code numbers 84/85 in elster 2019 (old line 52 in 2018) needs to be changed as it represents in 2019 more / other tax cases as in 2018. Finally we would have to change l10n_de_skr03 and l10n_de_skr04 as well, minimum the tax namings. 

It seems the adoption of just the line numbers in the tax statement config wizards view and the tax reports view would be the easy part to "fix". However if someone wants to print the 2018 values some lines could be missing. @astirpe Do you have any idea ? How do you adopt those changes in netherland (l10n_nl_vat_statement) ?

astirpe commented:
@tv-openbig 

For the changes regarding the line numbering and tax codes of version 2018 and version 2019, I could just allow the user to select the desired layout (2018, 2019, etc...) when creating/modifying the tax statement. The actual module can be reused, no need to create extra modules. This way, if someone needs to print an old version, it will be just re-printed as it was originally intended to be.

Technically I could just re-organize the actual python code of the module into multiple files, one for each version (eg.: `l10n_de_tax_statement_2018.py`, `l10n_de_tax_statement_2019.py`, etc...). This way the module will be easy to be extended in the future when a new version of text declaration will be published.

About merging or adding new lines, we should make sure that the calculations are done correctly, reflecting the changes made on the standard german localization modules (`l10n_de_skr03` and `l10n_de_skr04`). Do you already have the new german localization modules? Could you provide me with the code somehow?

tv-openbig commented:
@astirpe 

I was not aware about the line changes. Personally i do the mapping to the lines in elsteronline.de web portal by checking the tax code which is represented in the report lines in brackets.  The specific §13b tax cases where the change of lines appeared should be relatively seldom, therefore i didn't stumbled over it yet and most people should not have any issues to enter the right values in elsteronline.de portal. Anyway I will check the required changes in l10n_de_tax_statement.py this week. I assume some modifications in between the lines 265 - 442 and finally in the "number of the beast" line 666. ;-) 

To my opinion we also need a modified tag_ids field in the german account charts.

As an example. The following tax needs a change in the account chart, as the tag_ids entry "Ust-VA KZ 52"  should be "Ust-VA KZ 84" starting from 01/01/2019, if that tax will be used in an invoice in 2019. I am afraid we will be able to adopt that change to all existing invoices in 2019, if we would apply such a change not exactly at the date 01/01/2019 by also taking the risk to get wrong tax tags for the case we will create or modify invoices in the prior year. More worse we would get inconsistent reports for prior years if the tax tags are not saved at the time of invoice creation as a field in account.move.line.

<img width="1042" alt="Bildschirmfoto 2019-08-12 um 17 56 08" src="https://user-images.githubusercontent.com/8348598/62879682-fc8f3d80-bd2b-11e9-87cc-708bb96974e7.png">

Is it possible to adopt the potential issues i mentioned by modifying the l10n_de_tax_statement.py source code instead of changing the tag_ids ?

astirpe commented:
> Is it possible to adopt the potential issues i mentioned by modifying the l10n_de_tax_statement.py source code instead of changing the tag_ids ?

Yes, in my opinion this is the way to go :+1:

tv-openbig commented:
@astirpe Should we start here: https://github.com/OCA/l10n-germany/tree/11.0/l10n_de_tax_statement
I will make two copies from l10n_de_tax_statement.py:
l10n_de_tax_statement_2018.py (no change)
l10n_de_tax_statement_2019.py (changes according to official formular version for the year 2019).
Do you agree ?

astirpe committed:
Add version

astirpe commented:
@tv-openbig you could start here: https://github.com/onesteinbv/l10n-germany/tree/12_mig_l10n_de_tax_statement
I already split the `l10n_de_tax_statement_2018.py` and the `l10n_de_tax_statement_2019.py`. You could just modify the `l10n_de_tax_statement_2019.py`.

In the meanwhile I'm thinking about how to handle the config form.

astirpe committed:
Fix

tv-openbig committed:
adoption to UStVA Report 2019

The old lines 39 disappeared and was merged together with the old line
40 (now they are line 39 in 2019 edition). Furthermore the old lines 49
and 51 disappeared and was merged into the line 50. There is a
relevance for the tax codes (tag_ids). The old tag „USt-VA KZ 68“ have
to be mapped to the same line as the tag „USt-VA KZ 60“. The old tags
„USt-VA KZ 52“ and „USt-VA KZ 78“ have to be mapped the same way as the
old and single remaining „USt-VA 84“ (= base values). The old tags
„USt-VA KZ 53“ and „USt-VA KZ 79“ have to be mapped the same way as the
old and single remaining „USt-VA KZ 85“ (tax values). Furthermore we
might have formatting changes caused by the different line numbering.
Therefore i expect further adoptions in the files
„l10n_de_tax_statement_line.py“ and also in the
„l10n_de_tax_statement_config_wizard.xml“ and the
„l10n_de_tax_statement_config_wizard.py“. I assume we cannot avoid
deeper changes in order to adopt what we really want.

tv-openbig commented:
Hi @astirpe ,

would you please read my comment written as a commit notice. I don't think that we could avoid changing the tag_ids configuration (which tag is assigned to which tax) and the configuration wizards. For example the following tax tag is obsolete as it is now merged under another tax_tag (for example all taxes formerly married with tag_52_base should now be married with tag_84_base). 

Furthermore as a logical consequence the configuration wizard needs adoptions too 
(as some lines in the wizards view are obsolete too).

<img width="674" alt="Bildschirmfoto 2019-08-13 um 12 51 46" src="https://user-images.githubusercontent.com/8348598/62936266-a53faa80-bdc9-11e9-9315-d6a8e6c8ed24.png">

Did i missed something ?

astirpe committed:
Configuration of tags

astirpe commented on l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py:
```diff
@@ -282,39 +260,37 @@ def _finalize_lines_2019(lines):
+           - _56b - _57b - _58b - _59b
+    # calculate line 64
+    lines['64']['tax'] = lines['64']['tax'] * 1
+    _64b = lines['64']['tax'] + _60b + _62b + _63b
```
@tv-openbig  here _62b and _63b are not defined, it seems the definition was deleted. Could you check?

tv-openbig replied:
Correct, in next commit i will change that.

astirpe commented on l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py:
```diff
@@ -138,99 +132,87 @@
     },
-    '67': {
-        'code': '67', 'tax': 0.0,
+    '65': {
```
Is line 65 to be mapped with tag 39? If this is the case, the mapping is also missing in old version 2018.

tv-openbig replied:
This line is special and should be used only in December and the value should be entered manually. Therefore i assume in 2018 version it was line 67.

astirpe commented on l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py:
```diff
@@ -282,15 +282,65 @@ def _finalize_lines_2019(lines):
+
+def _get_tags_map_2019(config):
+    return {
+        config.tag_41_base.id: ('20', 'base'),
```
@tv-openbig could you double-check if the mapping tags-lines is correct?

tv-openbig replied:
I have deleted some lines, as we have some deleted lines in the offical VAT formular (elsteronline.de portal). However i assume that we have to assign other tags to some taxes (removal of tax tags "68" -> now "60" // "52" and "78" -> now "84" // "53" and "79" -> now "85"). All taxes with those tags needs to be revised (however just from 01/01/2019 not before !!!). How should we address that ?

astirpe replied:
@tv-openbig the lines you just removed https://github.com/OCA/l10n-germany/pull/39/commits/ba91812cb91b84434d8ca9320865a7f2ebf11ffa were instead doing the job:

- tag [68](https://github.com/OCA/l10n-germany/pull/39/commits/ba91812cb91b84434d8ca9320865a7f2ebf11ffa#diff-83ae6e5803d61484e5e11e55c1d1e4a1L323), mapping line 39 (same as tag 60)
- tags [52](https://github.com/OCA/l10n-germany/pull/39/commits/ba91812cb91b84434d8ca9320865a7f2ebf11ffa#diff-83ae6e5803d61484e5e11e55c1d1e4a1L329) and [78](https://github.com/OCA/l10n-germany/pull/39/commits/ba91812cb91b84434d8ca9320865a7f2ebf11ffa#diff-83ae6e5803d61484e5e11e55c1d1e4a1L333), mapping line 50 (same as tag 84)
- tags [53](https://github.com/OCA/l10n-germany/pull/39/commits/ba91812cb91b84434d8ca9320865a7f2ebf11ffa#diff-83ae6e5803d61484e5e11e55c1d1e4a1L330) and [79](https://github.com/OCA/l10n-germany/pull/39/commits/ba91812cb91b84434d8ca9320865a7f2ebf11ffa#diff-83ae6e5803d61484e5e11e55c1d1e4a1L334), mapping line 50 (same as tag 85)

Could you check?

tv-openbig replied:
I have again added the deleted lines.

astirpe commented:
@tv-openbig configuration of obsolete tags should be kept for retro-compatibility, in case someone is printing old 2018 statements.
Merged tags can be managed by calculating their sum in method `_finalize_lines_2019()`.

With latest commit I simplified the configuration form: now it doesn't depend on lines anymore; it's just a simple mapping of tags.

astirpe commented on l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py:
```diff
@@ -0,0 +1,346 @@
+    lines['56']['tax'] = lines['56']['tax'] * -1
+    _56b = lines['56']['tax']
+    lines['57']['tax'] = lines['57']['tax'] * -1
+    _57b = lines['59']['tax']
```
Should _57b be renamed to _59b?

tv-openbig replied:
No that was not my intention. I have corrected those values.

astirpe commented on l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py:
```diff
@@ -0,0 +1,346 @@
+    lines['57']['tax'] = lines['57']['tax'] * -1
+    _57b = lines['59']['tax']
+    lines['58']['tax'] = lines['58']['tax'] * -1
+    _58b = lines['60']['tax']
```
Should _58b be renamed to _60b?

tv-openbig replied:
Also an error from my side.

astirpe commented on l10n_de_tax_statement/models/l10n_de_tax_statement_2019.py:
```diff
@@ -0,0 +1,346 @@
+    _27b = lines['27']['tax']
+    _28b = lines['28']['tax']
+    _64b = lines['64']['tax']
+    _65b = lines['65']['tax']
```
_65b seems never used

tv-openbig replied:
That is the line i already explained above (to be manually entered)

astirpe committed:
Fix tests

tv-openbig committed:
changes of tax mappings to tags for the year 2019

tv-openbig committed:
tags mapping is still required

… compare conversation: https://github.com/OCA/l10n-germany/pull/39

astirpe committed:
Code review

astirpe commented:
@tv-openbig may I ask you your final review? After that, I will backport the latest developments to V11.

@haconsult Your issue should be fixed: this module now supports the old and the new statement format. Could you check?

astirpe committed:
Fix total + coverage

tv-openbig reviewed (commented):
Thanks @astirpe .I will make intense manual testing based on V11, after your backport.

tv-openbig commented:
@astirpe May i merge your pull request ?

astirpe commented:
Yes, I don't have other changes to make. Thanks!

haconsult commented:
Works, thank you!

haconsult commented:
I don't know if this is the right place, however there is a new issue.
When I created more than one Tax Report I can not go back to the tax report menu. It produces the following issue:

Fehler:
Odoo Server Error

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/api.py", line 1039, in get
    value = self._data[key][field][record._ids[0]]
KeyError: 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/fields.py", line 981, in __get__
    value = record.env.cache.get(record, self)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/api.py", line 1041, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: ('l10n.de.tax.statement(1,).tax_total', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/models.py", line 4718, in ensure_one
    _id, = self._ids
ValueError: too many values to unpack (expected 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 656, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 314, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 698, in dispatch
    result = self._call_function(**self.params)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 346, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/http.py", line 519, in response_wrap
    response = f(*args, **kw)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/addons/web/controllers/main.py", line 904, in search_read
    return self.do_search_read(model, fields, offset, limit, domain, sort)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/addons/web/controllers/main.py", line 926, in do_search_read
    offset=offset or 0, limit=limit or False, order=sort or False)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/models.py", line 4589, in search_read
    result = records.read(fields)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/models.py", line 2791, in read
    vals[name] = convert(record[name], record, use_name_get)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/models.py", line 5117, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/fields.py", line 985, in __get__
    self.determine_value(record)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/fields.py", line 1098, in determine_value
    self.compute_value(recs)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/fields.py", line 1052, in compute_value
    self._compute_value(records)
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/fields.py", line 1043, in _compute_value
    getattr(records, self.compute)()
  File "/usr/local/bin/custom-addons/l10n_de_tax_statement/models/l10n_de_tax_statement.py", line 370, in _compute_tax_total
    if self.version == '2019':
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/fields.py", line 979, in __get__
    record.ensure_one()
  File "/usr/lib/python3.6/site-packages/odoo-12.0.post20190723-py3.6.egg/odoo/models.py", line 4721, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: l10n.de.tax.statement(1, 2, 3)

astirpe commented:
Thanks for reporting the error. I will fix it tomorrow.

astirpe commented:
@haconsult fixed in #58 and #59

haconsult commented:
Works great, thank you!
